### PR TITLE
0.1.x handle arakoon errors

### PR DESCRIPTION
**Pull request by JeffreyDevloo**

### CODE REVIEW

File: src/db/arakoon/pyrakoon/client.py [hunk 1]
```diff
@@ -50,6 +53,64 @@ def new_function(self, *args, **kw):
...
+                while time.time() < deadline:
+                    try:
+                        result = f(self, *args, **kwargs)
+                        duration = time.time() - start
+                        if duration > max_duration:
+                            logger.warning('Pyrakoon call {0} took {1}s'.format(f.__name__, round(duration, 2)))
+                        return result
+                    except (ArakoonNoMaster, ArakoonNodeNotMaster, ArakoonSocketException, ArakoonNotConnected, ArakoonGoingDown) as ex:
+                        # (ArakoonSockNotReadable, ArakoonSockReadNoBytes, ArakoonSockSendError)  are the socket exception that can be retried
+                        if not is_read_only and isinstance(ex, (ArakoonSocketException, ArakoonGoingDown)) and not isinstance(ex, (ArakoonSockNotReadable, ArakoonSockReadNoBytes, ArakoonSockSendError)):
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client.py]: line too long

File: src/db/arakoon/pyrakoon/client.py [hunk 2]
```diff
@@ -64,184 +125,265 @@ class PyrakoonClient(object):
...
-                                        self._client.multiGet if must_exist is True else self._client.multiGetOption,
-                                        keys):
+        :param keys: All keys to fetch
+        :type keys" List[str]
+        :param must_exist: Should all listed keys exist
+        :type must_exist: bool
+        :return: Generator that yields key value pairs
+        :rtype: iterable[Tuple[str, any]
+        """
+        func = self._client.multiGet if must_exist is True else self._client.multiGetOption
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client.py]: `if .... is True`
is this a type check in disguise ?

File: src/db/arakoon/pyrakoon/client.py [hunk 3]
```diff
@@ -50,6 +53,64 @@ def new_function(self, *args, **kw):
...
+                            logger.warning('Pyrakoon call {0} took {1}s'.format(f.__name__, round(duration, 2)))
+                        return result
+                    except (ArakoonNoMaster, ArakoonNodeNotMaster, ArakoonSocketException, ArakoonNotConnected, ArakoonGoingDown) as ex:
+                        # (ArakoonSockNotReadable, ArakoonSockReadNoBytes, ArakoonSockSendError)  are the socket exception that can be retried
+                        if not is_read_only and isinstance(ex, (ArakoonSocketException, ArakoonGoingDown)) and not isinstance(ex, (ArakoonSockNotReadable, ArakoonSockReadNoBytes, ArakoonSockSendError)):
+                            raise
+                        # Drop all master connections and master related information
+                        self._client._client_masterId = None
+                        self._client.dropConnections()
+                        sleep_time = back_off_period * tries
```
Comment (toolslive) [inline src/db/arakoon/pyrakoon/client.py]: why linear iso exponential?

File: src/db/arakoon/pyrakoon/client.py (hunk 2 above)
Comment (JeffreyDevloo) [reply src/db/arakoon/pyrakoon/client.py]: It is!

File: src/db/arakoon/pyrakoon/client.py (hunk 1 above)
Comment (JeffreyDevloo) [reply src/db/arakoon/pyrakoon/client.py]: Damn you PEP8

File: src/db/arakoon/pyrakoon/client.py [hunk 4]
```diff
@@ -48,6 +51,66 @@ def new_function(self, *args, **kw):
     return wrap
 
 
+def handle_arakoon_errors(is_read_only=False, max_duration=0.5, override_retry=False):
+    # type: (bool, float) -> Any
```
Comment (simonOpenV) [inline src/db/arakoon/pyrakoon/client.py]: missing parameter entry of 'override_retry'

File: src/db/arakoon/pyrakoon/client.py [hunk 5]
```diff
@@ -57,190 +120,270 @@ class NoLockAvailableException(Exception):
...
-    Arakoon client wrapper:
-    * Uses json serialisation
-    * Raises generic exception
+    Arakoon client wrapper
     """
     _logger = Logger('extensions')
 
-    def __init__(self, cluster, nodes):
+    def __init__(self, cluster, nodes, retries=10, retry_back_off_multiplier=2, retry_interval_sec=2):
+        # type: (str, Dict[str, Tuple[str, int]], float, float) -> None
```
Comment (simonOpenV) [inline src/db/arakoon/pyrakoon/client.py]: missing 1 parameter

**Pull request merged**